### ISSUE TITLE
CA1849: Do not report a diagnostic if there is a check for task is resolved

### ISSUE THREAD

**Issue by mahdiva** [dotnet/roslyn-analyzers#5416]
### Analyzer

**Diagnostic ID**: [CA1849](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1849): Call async methods when in an async method

### Describe the bug

Currently CA1849 does not check whether there is an if condition checking if Task is resolved, when using a sync method/property, such as `Task<T>.Result`.

This includes checks using `Task.IsCompleted`, `Task.IsCompletedSuccessfully`, `Task.IsFaulted`, `Task.IsCancelled` etc...

### Steps To Reproduce

```csharp
using System.Threading.Tasks;

class Test
{
    Task T()
    {
        Task<int> t = null;
        int result = 0;
        if (t.IsCompleted)
        {
            result = t.Result;
        }
        return Task.FromResult(result);
    }
}
```

### Expected behavior

Since `t.Result` is guarded by the `t.IsCompleted` check, there should be no diagnostic reported for this case

**Comment by xPaw** [comment]
This diagnostic is also triggered when tasks are awaited with `Task.WhenAll`, and after `task.Status == TaskStatus.RanToCompletion` checks.

**Comment by ayeltsov** [comment]
One more case, accessing Result in task continuations.

```csharp
            // Problem does reproduce here
            _ = Task.FromResult(true).ContinueWith(t =>
            {
                _ = t.Result; // CA1849 on t.Result. Am I missing anything here? Pretty sure it's analyzer who's wrong, not the code.
                return Task.CompletedTask; // continuation has to return Task to repro the issue
            }).Unwrap(); // Unrap or not, still reproes

            // Counter-example - no repro here
            _ = Task.FromResult(true).ContinueWith(t =>
            {
                _ = t.Result; // No CA1849
                return;
            });

```
